### PR TITLE
Add support for custom install-configs for clusterpools (supported in Hive 1.0.16+/ACM 2.2+)

### PR DESCRIPTION
**Pull request by gurnben**
## Summary of Changes

This PR now allows the user to create an install-config (and we'll create the secret for them), bring their own install config template (and we can create the secret for them) or select a pre-existing secret.  

This change might necessitate documentation/readme changes - especially on required vars for full automation

### CODE REVIEW

File: clusterpools/apply.sh
Comment (dhaiducek) [inline clusterpools/apply.sh]: Maybe this is just me, but calling this a "secret" is confusing to me since in my mind configurations are generally not secrets, and I started wondering if it needed my OCP pull secret.

I propose leaving off the word "secret" in this section except perhaps to make it clear that the resource is being stored as a secret.

Comment (dhaiducek) [inline clusterpools/apply.sh]: automatially --> automatically

File: clusterpools/apply.sh
Comment (dhaiducek) [inline clusterpools/apply.sh:613]: Why do you have both `CLUSTERPOOL_INSTALL_CONFIG_SECRET` and `CLUSTERPOOL_INSTALL_CONFIG_SECRET_NAME`?
I'm not immediately seeing a reason to have them be different names...

File: clusterpools/apply.sh
Comment (dhaiducek) [inline clusterpools/apply.sh:610]: Wait. If I give a custom name, you're tagging a random identifier on it? That's not correct, is it?

File: clusterpools/apply.sh
Comment (dhaiducek) [inline clusterpools/apply.sh:192]: I'd love for this to fill in the base domain if I've already given `CLUSTERPOOL_AWS_BASE_DOMAIN` and fill in the name from `CLUSTERPOOL_INSTALL_CONFIG_SECRET_NAME`.

**Review by dhaiducek — COMMENTED** [review]
Looks really good. I have a bunch of things to say. 😄

File: clusterpools/templates/install-config.aws.yaml.template
Comment (dhaiducek) [inline clusterpools/templates/install-config.aws.yaml.template:38]: it's --> its

File: clusterpools/templates/install-config.azure.yaml.template
Comment (dhaiducek) [inline clusterpools/templates/install-config.azure.yaml.template:39]: it's --> its

File: clusterpools/templates/install-config.gcp.yaml.template
Comment (dhaiducek) [inline clusterpools/templates/install-config.gcp.yaml.template:31]: it's --> its

File: clusterpools/apply.sh
Comment (gurnben) [reply clusterpools/apply.sh]: @dhaiducek Ahh yes, so I named the input var with the `_SECRET_NAME` at the end because we specifically want the user to provide us the _name_ of install config secret on the cluster that they want us to use to override this.  This variable otherwise holds the generated name of the secret we'll create on the cluster and point hive to.  Its confusing to me primarily because hive uses secrets (as opposed to configmaps) to hold this info!

I would propose we keep the secret naming to keep the resource type obvious!

File: clusterpools/apply.sh
Comment (gurnben) [reply clusterpools/apply.sh:613]: Yepp - so the difference here is that one of them points to the file that the user has on their system that contains the yaml that should be put into a secret on the cluster.  The latter is the name of a secret that they already created that contains the install-config on the cluster that they want us to reference (and skip creating another secret).

File: clusterpools/apply.sh
Comment (gurnben) [reply clusterpools/apply.sh:610]: I'm appending a random identifier to the secret name in this case, not the clusterpool.  In this specific case, the user wants us to create an install-config-secret for them for a clusterpool that has a custom name.  If we create the secret without a random identifier and they try to re-create the same pool (after delete) without deleting the secret, or if they try to update the pool by rerunning, the process will fail because secrets can't be edited.  Adding the random identifier prevents secret collisions.

File: clusterpools/templates/install-config.aws.yaml.template
Comment (gurnben) [reply clusterpools/templates/install-config.aws.yaml.template:38]: @dhaiducek I don't know how to say it but I stole this install-config from the RHACM product's UI page so we have it wrong in the product!

File: clusterpools/templates/install-config.azure.yaml.template
Comment (gurnben) [reply clusterpools/templates/install-config.azure.yaml.template:39]: @dhaiducek I don't know how to say it but I stole this install-config from the RHACM product's UI page so we have it wrong in the product!

File: clusterpools/templates/install-config.gcp.yaml.template
Comment (gurnben) [reply clusterpools/templates/install-config.gcp.yaml.template:31]: @dhaiducek I don't know how to say it but I stole this install-config from the RHACM product's UI page so we have it wrong in the product!

File: clusterpools/apply.sh
Comment (dhaiducek) [reply clusterpools/apply.sh:613]: Aren't you referring to `CLUSTERPOOL_INSTALL_CONFIG_FILE`? I'm talking about the other two variable names.

File: clusterpools/apply.sh
Comment (dhaiducek) [reply clusterpools/apply.sh]: I agree about the variable name. I'm talking about the actual CLI output:
Maybe remove some of the "secret" references in here. I think mainly this one (since we're already listing secrets and we've referred to them as secrets previously):
```
Create a new Install Config (interactive).
Enter the number corresponding to your desired Install Config from the list above
```
or:
```
Create a new Install Config (interactive).
Enter the number corresponding to your desired secret containing the Install Config from the list above
```

File: clusterpools/apply.sh
Comment (dhaiducek) [reply clusterpools/apply.sh:610]: Ah, that makes sense. Because if it's not custom, then it already has the identifier.
...but secrets can be updated, right? I feel like this goes against other flows where we overwrite ClusterClaims with reckless abandon.

File: clusterpools/templates/install-config.aws.yaml.template
Comment (dhaiducek) [reply clusterpools/templates/install-config.aws.yaml.template:38]: YIKES!!!

File: clusterpools/templates/install-config.azure.yaml.template
Comment (dhaiducek) [reply clusterpools/templates/install-config.azure.yaml.template:39]: PANIC!!!

File: clusterpools/templates/install-config.gcp.yaml.template
Comment (dhaiducek) [reply clusterpools/templates/install-config.gcp.yaml.template:31]: EVERYTHING'S BROKEN!!!

File: clusterpools/apply.sh
Comment (gurnben) [reply clusterpools/apply.sh:610]: @dhaiducek well.... secrets can't _exactly_ be updated.  Its tricky to be honest.  Re-creating was much, much easier

File: clusterpools/apply.sh
Comment (gurnben) [reply clusterpools/apply.sh]: Ahhh, good point there - let me clear that up!

Comment (gurnben) [reply clusterpools/apply.sh]: Oh oh oh - @dhaiducek this line:
```
printf "${BLUE}- note: to skip this step in the future and create a new install-config secret from a file automatially, export CLUSTERPOOL_INSTALL_CONFIG_FILE${CLEAR}\n"
```
is telling the user that if they set CLUSTERPOOL_INSTALL_CONFIG_FILE, we'll read in the contents of that file and create and use a secret without their input!

Comment (gurnben) [reply clusterpools/apply.sh]: Commit incoming shortly!

File: clusterpools/apply.sh
Comment (gurnben) [reply clusterpools/apply.sh:613]: Ok, now that I'm sound of mind and not multitasking - what is happening here is:
```
    if [[ "$CLUSTERPOOL_INSTALL_CONFIG_FILE" == "" ]]; then
        printf "${BLUE}- note: to skip this step in the future and create a new install-config secret from a file automatically, export CLUSTERPOOL_INSTALL_CONFIG_FILE${CLEAR}\n"
        if [[ "$CLUSTERPOOL_INSTALL_CONFIG_SECRET" == "" ]]; then
            printf "${BLUE}- note: to skip this step in the future and use a pre-existing install-config secret, export CLUSTERPOOL_INSTALL_CONFIG_SECRET${CLEAR}\n"
```

If the user tells us "here is a file that contains an install config (CLUSTERPOOL_INSTALL_CONFIG_FILE), I would like you to create a secret from it and use that", otherwise if a user says "here is a secret I already have on the cluster (maybe from a previous run) that I would like you to use as the install config", otherwise we prompt them.  How's that sound?

File: clusterpools/apply.sh
Comment (gurnben) [reply clusterpools/apply.sh:192]: @dhaiducek are you meaning the install-config entry for CLUSTERPOOL_AWS_BASE_DOMAIN?  If so - that value is actually auto-filled-in by hive at runtime, you're supposed to leave that empty!

File: clusterpools/apply.sh
Comment (dhaiducek) [reply clusterpools/apply.sh:613]: I was confused about why we have both `CLUSTERPOOL_INSTALL_CONFIG_SECRET` and `CLUSTERPOOL_INSTALL_CONFIG_SECRET_NAME` and whether if I set the latter if anything would break or if both were really needed.

But `CLUSTERPOOL_INSTALL_CONFIG_SECRET_NAME` is used as a placeholder and would be overwritten here regardless: https://github.com/open-cluster-management/lifeguard/blob/7ffef5b6752115039c246acb6fd7259f466abbb1/clusterpools/apply.sh#L606-L610

And then `CLUSTERPOOL_INSTALL_CONFIG_SECRET` is the official variable used to set the secret name.

I think I get it now. 🙂

File: clusterpools/apply.sh
Comment (dhaiducek) [reply clusterpools/apply.sh:192]: Well okay, then! So should we put a brief comment so people like me don't wonder?
```
metadata:
  name: 
baseDomain: # skip, hive will inject it from the ClusterPool
```
Or leave it off entirely?

File: clusterpools/apply.sh
Comment (dhaiducek) [reply clusterpools/apply.sh]: Right--that's the line that I'm thinking it's okay to leave alone. It's the others inside of the creation portion.

File: clusterpools/apply.sh
Comment (dhaiducek) [inline clusterpools/apply.sh:628]: Should this be "N"? We don't want to exit if they want to continue, do we?

File: clusterpools/apply.sh
Comment (dhaiducek) [inline clusterpools/apply.sh]: and --> an

**Review by dhaiducek — DISMISSED** [review]
At this point, my comments don't really affect functionality (except that latest continue or not thing--that might need fixing). Approving--I'll let you decide what's best. 😄

File: clusterpools/apply.sh
Comment (dhaiducek) [reply clusterpools/apply.sh]: But it was an initial confusion I had that I no longer have, and others may not have the same misunderstanding I did.

File: clusterpools/apply.sh
Comment (gurnben) [reply clusterpools/apply.sh:192]: Aha - again I copied straight from ACM 😆

File: clusterpools/apply.sh
Comment (gurnben) [reply clusterpools/apply.sh:628]: Ahh - the logic here is "do you wish to continue" and if the answer is _not_ "Y" or "y" we bail

File: clusterpools/apply.sh
Comment (gurnben) [reply clusterpools/apply.sh:192]: I'm a bit concerned that if we add a comment there we might break hive's injection 😨.  Let me merge this and then try it with a comment later!

File: clusterpools/apply.sh
Comment (gurnben) [reply clusterpools/apply.sh]: Done!

File: clusterpools/apply.sh
Comment (dhaiducek) [reply clusterpools/apply.sh:628]: Oh yeah--that makes sense. Didn't notice the negation 🙂

**Review by dhaiducek — APPROVED** [review]

**Pull request merged**
